### PR TITLE
Move show blame and open on the code host actions to the actions sidebar

### LINKED ISSUE #41522: Regression: `OpenInEditorActionItem` can cause infinite request retries on error
More context: https://sourcegraph.slack.com/archives/CMBA8F926/p1662559931542359

### Steps to reproduce

1. Have invalid settings (because schema has been changed, for example)
2. Open a blob view, like: https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/Layout.tsx?L125
3. See your requests go up:
screenshot_2022-09-07_16 32 24
4. Browser tab crashes and you get blocked by Cloudflare

### Debug info

I nailed this down to this PR https://github.com/sourcegraph/sourcegraph/pull/41105 which introduced these lines https://github.com/sourcegraph/sourcegraph/blob/b7b9d52c2e7b15b04e6faa01e4fd23137b74c570/client/web/src/open-in-editor/OpenInEditorActionItem.tsx#L54-L66

they show up in the stack trace of the 10s of thousands of `OverwriteSettings` calls that seem to happen:

image

Here's the error response

image

/cc @muratsu @jjinnii @ryankscott

### PR DESCRIPTION
**Pull request by taras-yemets**
Closes https://github.com/sourcegraph/sourcegraph/issues/41327
Closes #41522

1. Moves git blame to the actions (aka extensions) sidebar.
2. Move open on the code host icons to the actions (aka extensions) sidebar on all views except for compare and commit.
3. Refactors `SimpleActionItem` component to conditionally render buttons and links and accept different types of children (icons in this case).
4. Hides actions (aka extensions) sidebar on the commit and compare views if `enableLegacyExtensions` site config parameter is `false`.
5. Renders git blame and open in editor actions only in the blob view.

<details>
<summary><b>Changes table</b></summary>
<table>
<thead>
<tr>
<td>Page</td>
<td>Before</td>
<td>After</td>
<td>Comment</td>
</tr>
</thead>
<tbody>
<tr>
<td>Repo page</td>
<td>
Screenshot 2022-09-09 at 08 14 54
</td>
<td>
<code>enableLegacyExtensions: true</code>(Codecov icon is rendered) Screenshot 2022-09-09 at 08 15 18
</td>
<td>Click navigates to <a href="https://github.com/sourcegraph/sourcegraph">sourcegraph/sourcegraph</a></td>
</tr>
<tr>
<td>Repo commits page</td>
<td>
Screenshot 2022-09-09 at 08 19 11
</td>
<td>Not changed</td>
<td>Click navigates to <a href="https://github.com/sourcegraph/sourcegraph">sourcegraph/sourcegraph</a>. It doesn't look relevant to me, ⚠️ <b>is it a bug?</b> Should it navigate to the repo commits page on the code host?</td>
</tr>
<tr>
<td>Repo commit page</td>
<td>
Screenshot 2022-09-09 at 08 22 32
</td>
<td>
1. <code>enableLegacyExtensions: false</code>Screenshot 2022-09-09 at 08 34 48
2. <code>enableLegacyExtensions: true</code>(no extensions active on the screenshot, but there could be some)Screenshot 2022-09-09 at 08 48 54
</td>
<td></td>
</tr>
<tr>
<td>Repo branches page</td>
<td>
Screenshot 2022-09-09 at 08 23 41
</td>
<td>Not changed</td>
<td>Click navigates to <a href="https://github.com/sourcegraph/sourcegraph">sourcegraph/sourcegraph</a>. It doesn't look relevant to me, ⚠️ <b>is it a bug?</b> Should it navigate to the repo branches page on the code host?</td>
</tr>
<tr>
<td>Repo tags page</td>
<td>
Screenshot 2022-09-09 at 08 24 27
</td>
<td>Not changed</td>
<td>Click navigates to <a href="https://github.com/sourcegraph/sourcegraph">sourcegraph/sourcegraph</a>. It doesn't look relevant to me, ⚠️ <b>is it a bug?</b></td>
</tr>
<tr>
<td>Repo compare page</td>
<td>
Screenshot 2022-09-09 at 08 25 59
</td>
<td>Not changed</td>
<td>Click navigates to <a href="https://github.com/sourcegraph/sourcegraph">sourcegraph/sourcegraph</a>. It doesn't look relevant to me, ⚠️ <b>is it a bug?</b> Should it navigate to the compare page on the code host?</td>
</tr>
<tr>
<td>Repo compare page (branches to compare selected)</td>
<td>
1. <code>enableLegacyExtensions: false</code>Screenshot 2022-09-09 at 08 27 25
2. <code>enableLegacyExtensions: true</code>(no extensions active on the screenshot, but there could be some)
Screenshot 2022-09-09 at 08 51 56
</td>
<td>
Screenshot 2022-09-09 at 08 36 12
</td>
<td></td>
</tr>
<tr>
<td>Repo contributors page</td>
<td>
Screenshot 2022-09-09 at 08 28 49
</td>
<td>Not changed</td>
<td>Click navigates to <a href="https://github.com/sourcegraph/sourcegraph">sourcegraph/sourcegraph</a>. It doesn't look relevant to me, ⚠️ <b>is it a bug?</b></td>
</tr>
<tr>
<td>Repo code graph page</td>
<td>
Screenshot 2022-09-09 at 08 29 35
</td>
<td>Not changed</td>
<td>Click navigates to <a href="https://github.com/sourcegraph/sourcegraph">sourcegraph/sourcegraph</a>. It doesn't look relevant to me, ⚠️ <b>is it a bug?</b> Do we need go to the code host action on this page?</td>
</tr>
<tr>
<td>Repo settings page</td>
<td>
Screenshot 2022-09-09 at 08 30 49
</td>
<td>Not changed</td>
<td>Click navigates to <a href="https://github.com/sourcegraph/sourcegraph">sourcegraph/sourcegraph</a>. It doesn't look relevant to me, ⚠️ <b>is it a bug?</b> Do we need go to the code host action on this page?</td>
</tr>
<tr>
<td>Repo folder page</td>
<td>
Screenshot 2022-09-09 at 08 31 44
</td>
<td>
Screenshot 2022-09-09 at 08 40 00
</td>
<td>Both screenshots don't have a view on the code host action. The corresponding page (folder page) likely exists on the code host. ⚠️ <b>Is it a bug?</b></td>
</tr>
<tr>
<td>Repo blob page</td>
<td>
Screenshot 2022-09-09 at 08 41 57
</td>
<td>
Screenshot 2022-09-09 at 08 42 37
</td>
<td>Git blame and open in editor actions are rendered only in the blob view. Previously they were rendered on the tree and a few other views.</td>
</tr>
<tr>
</tbody>
</table>
</details>

## Test plan
- ensure git extras (blame) icon is rendered in the actions (aka extensions) sidebar in the tree view
- ensure view on the code host icon is rendered in the actions (aka extensions) sidebar on all views except for the commit and compare ones
- ensure actions (aka extensions) sidebar is not rendered in the commit and compare views if `enableLegacyExtensions` site config parameter is `false`

## App preview:

- [Web](https://sg-web-taras-yemets-move-integrations.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-epfhlryeyr.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/web/src/open-in-editor/OpenInEditorActionItem.tsx [hunk 1]
```diff
@@ -48,31 +47,29 @@ export const OpenInEditorActionItem: React.FunctionComponent<OpenInEditorActionI
...
-        )
+                setSettingsCascadeOrError(settings)
+            }
+        })
 
         return () => {
             settingSubscription?.unsubscribe()
         }
-    }, [settingSubscription, props.platformContext.settings, props.platformContext])
+    }, [props.platformContext])
```
Comment (taras-yemets) [inline client/web/src/open-in-editor/OpenInEditorActionItem.tsx:72]: Calling `setSettingSubscription` in `useEffect` which had `settingSubscription` in its dependencies caused an infinite loop after I refactored the way we use `SimpleActionItem` component.
Why do we need to have `settingSubscription` in the state? It seems we don't use it elsewhere. 
cc: @vdavid

File: client/web/src/repo/RepoContainer.tsx [hunk 2]
```diff
@@ -520,3 +524,11 @@ function getIsCodeIntelRepositoryBadgeVisible(options: {
...
     return isCodeIntelRepositoryBadgeEnabled && isCodeIntelRepositoryBadgeVisibleOnRoute
 }
+
+// Show go to the code host action only on commit and compare views. On the other views it's rendered in the actions sidebar.
+function getIsGoToCodeHostActionVisible(match: RepoContainerProps['match']): boolean {
+    return Boolean(
+        matchPath(match.url, { path: match.path + commitRevSpecPath }) ||
+            matchPath(match.url, { path: match.path + compareSpecPath })
+    )
+}
```
Comment (taras-yemets) [inline client/web/src/repo/RepoContainer.tsx:534]: Is there a better way to define whether we are on either commit or compare view?
Should we move rendering `GoToCodeHostAction` into the `RepoHeaderContributionPortal` logic to _client/web/src/repo/routes.tsx_?

File: client/web/src/extensions/components/ActionItemsBar.tsx [hunk 3]
```diff
@@ -185,8 +192,11 @@ const actionItemClassName = classNames(
  * TODO: description
```
Comment (philipp-spiess) [inline client/web/src/extensions/components/ActionItemsBar.tsx:192]: 🤔

File: client/web/src/open-in-editor/OpenInEditorActionItem.tsx [hunk 4]
```diff
@@ -48,31 +47,29 @@ export const OpenInEditorActionItem: React.FunctionComponent<OpenInEditorActionI
...
-                            .updateSettings(subject.subject.id, JSON.stringify(migratedSettings, null, 4))
-                            .then(() => {
-                                console.log('Migrated items successfully.')
-                            })
-                            .catch(() => {
-                                // TODO: Update failed, handle this later
-                            })
-                    }
-                    setSettingsCascadeOrError(settings)
+        const settingSubscription = props.platformContext.settings.subscribe(settings => {
```
Comment (philipp-spiess) [inline client/web/src/open-in-editor/OpenInEditorActionItem.tsx:50]: Let's revert this so we can properly rebase it against https://github.com/sourcegraph/sourcegraph/pull/41527

**Review by philipp-spiess — APPROVED** [review]
Cookie Monster

File: client/web/src/extensions/components/ActionItemsBar.tsx [hunk 5]
```diff
@@ -224,7 +234,29 @@ export const ActionItemsBar = React.memo<ActionItemsBarProps>(function ActionIte
                         <Icon aria-hidden={true} svgPath={mdiMenuUp} />
                     </Button>
                 )}
-                <OpenInEditorActionItem platformContext={props.platformContext} />
+
+                {source !== 'compare' && source !== 'commit' && (
```
[the remark this answers is not in the document]
Comment (taras-yemets) [reply client/web/src/extensions/components/ActionItemsBar.tsx:238]: We render `ActionItemsBar` (and probably expect the go to the code host action to be present) on all [these routes](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@89a24fe822c3cc8f3c714ec512da357d9b689827/-/blob/client/web/src/repo/routes.tsx?L144%3A5-176%3A9=), but we want to render git blame and open in editor actions only in the blob view.
Does it sound correct to you?

File: client/web/src/open-in-editor/OpenInEditorActionItem.tsx [hunk 6]
```diff
@@ -99,28 +96,34 @@ export const OpenInEditorActionItem: React.FunctionComponent<OpenInEditorActionI
                         <SimpleActionItem
                             key={editor.id}
                             tooltip={`Open file in ${editor?.name}`}
-                            iconURL={`${assetsRoot}/img/editors/${editor.id}.svg`}
-                            onClick={() => {
+                            onSelect={() => {
```
[the remark this answers is not in the document]
Comment (taras-yemets) [reply client/web/src/open-in-editor/OpenInEditorActionItem.tsx:99]: We pass an event handler to the underlying component (`ButtonLink`) [code](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@89a24fe822c3cc8f3c714ec512da357d9b689827/-/blob/client/wildcard/src/components/ButtonLink/ButtonLink.tsx?L83%3A4-92%3A6=).
What do you think?

File: client/web/src/repo/actions/GoToCodeHostAction.tsx [hunk 7]
```diff
@@ -36,6 +33,8 @@ interface Props extends RevisionSpec, Partial<FileSpec> {
     fetchFileExternalLinks: typeof fetchFileExternalLinks
 
     actionType?: 'nav' | 'dropdown'
+
+    as?: 'repoHeaderAction' | 'actionItemBarAction'
```
[the remark this answers is not in the document]
Comment (taras-yemets) [reply client/web/src/repo/actions/GoToCodeHostAction.tsx:37]: Makes sense. I will look for a more suitable name. Thank you!

File: client/web/src/extensions/components/ActionItemsBar.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (philipp-spiess) [reply client/web/src/extensions/components/ActionItemsBar.tsx:238]: Ohh okay gotcha! 🙂

File: client/web/src/repo/actions/GoToCodeHostAction.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (taras-yemets) [reply client/web/src/repo/actions/GoToCodeHostAction.tsx]: Addressed in https://github.com/sourcegraph/sourcegraph/pull/41431/commits/bf7992775c9954d752df22512411c74a1db5d100

File: client/web/src/open-in-editor/OpenInEditorActionItem.tsx (hunk 1 above)
Comment (taras-yemets) [reply client/web/src/open-in-editor/OpenInEditorActionItem.tsx]: Addressed in https://github.com/sourcegraph/sourcegraph/pull/41527

File: client/web/src/extensions/components/ActionItemsBar.tsx (hunk 3 above)
Comment (taras-yemets) [reply client/web/src/extensions/components/ActionItemsBar.tsx]: Addressed in https://github.com/sourcegraph/sourcegraph/pull/41431/commits/3e14c2508d7516fb1aed4e44643e9dc67283ae17

**Comment by taras-yemets** [comment]
Updated the PR description with the changes table.
cc: @philipp-spiess, @vdavid, @jjinnii

**Pull request merged**